elastic/beats
# Add deployment name in pod's meta

## What does this PR do?
This PR adds `kubernetes.deployment.name` in Pods' metadata. Right now we only add `kubernetes.replicaset.name` when a Pod is handled by a Deployment since a Deployment controls the Pod through a ReplicaSet and the ReplicaSet is considered as the direct owner of the Pod in k8s API. In this we need an extra step check to retrieve the owner of the ReplicaSet if possible.


## Why is it important?
So as to correlate Pods with Deployments controlling them.


## How to test this PR locally
1. Deploy a sample Deployment on k8s: `kubectl apply -f https://k8s.io/examples/controllers/nginx-deployment.yaml `
2. Deploy Metricbeat ([instructions](https://github.com/elastic/observability-dev/blob/master/docs/dc/testing/k8s.md)) on k8s and verify that events shiped by `pod` and `state_pod` metricset include `kubernetes.deployment.name` when `kubernetes.replicaset.name` exists. As an example see screenshots below.

Extra: Verify that `add_kubernetes_metadata` adds `kubernetes.deployment.name` when `kubernetes.replicaset.name` exists, using the following configuration with Filebeat on k8s:
```
filebeat.inputs:
- type: container
  paths:
    - /var/log/containers/*.log
  processors:
    - add_kubernetes_metadata:
        host: ${NODE_NAME}
        matchers:
        - logs_path:
            logs_path: "/var/log/containers/"
```

## Related issues
- Closes https://github.com/elastic/beats/issues/20991


## Screenshots
![Screenshot 2021-01-21 at 11 29 37 AM](https://user-images.githubusercontent.com/11754898/105334202-447fe880-5bdf-11eb-8118-affe0b50d699.png)
![Screenshot 2021-01-21 at 11 06 15 AM](https://user-images.githubusercontent.com/11754898/105334207-46e24280-5bdf-11eb-8092-e9c8bdf36485.png)
